# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1251789993 -7200
#      Tue Sep 01 09:26:33 2009 +0200
# Branch release-3-2-x
# Node ID f3e70f64c30c4bb52d8ff3ffb4676477720140ee
# Parent  279d2f4102d721ff3204f661da88423848eda64d
std.m: correctly work along singleton dimension

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,13 @@
+2009-08-06  John W. Eaton  <jwe@octave.org>
+
+	* statistics/base/std.m: Correctly work along singleton dimension.
+	From Christoph Ellenberger <C.Ellenberger@gmx.net>.
+
 2009-08-06  Olaf Till <olaf.till@uni-jena.de>
 
 	* geometry/griddata.m: Linearize arrays.
 
 2009-08-25  John W. Eaton  <jwe@octave.org>
 
 	* plot/__gnuplot_open_stream__.m: Save pid in __plot_stream__ property.
 	* plot/gnuplot_drawnow.m: Wait for gnuplot subprocess when printing.
diff --git a/scripts/statistics/base/std.m b/scripts/statistics/base/std.m
--- a/scripts/statistics/base/std.m
+++ b/scripts/statistics/base/std.m
@@ -59,38 +59,39 @@
 
 ## Author: jwe
 
 function retval = std (a, opt, dim)
 
   if (nargin < 1 || nargin > 3)
     print_usage ();
   endif
-  if nargin < 3
+  if (nargin < 3)
     dim = find (size (a) > 1, 1);
     if (isempty (dim))
       dim = 1;
     endif
   endif
   if (nargin < 2 || isempty (opt))
     opt = 0;
   endif
 
   n = size (a, dim);
   if (n == 1)
-    retval = zeros (sz);
+    retval = zeros (size (a));
   elseif (numel (a) > 0)
     retval = sqrt (sumsq (center (a, dim), dim) / (n + opt - 1));
   else
     error ("std: x must not be empty");
   endif
 
 endfunction
 
 %!test
 %! x = ones (10, 2);
 %! y = [1, 3];
 %! assert(std (x) == [0, 0] && abs (std (y) - sqrt (2)) < sqrt (eps));
+%! assert (std (x, 0, 3), zeros (10, 2))
+%! assert (std (ones (3, 1, 2), 0, 2), zeros (3, 1, 2))
 
 %!error std ();
 
 %!error std (1, 2, 3, 4);
-
